products/marketing_analytics/backend/hogql_queries/__snapshots__/test_conversion_goal_processor.ambr

```diff
@@ -1,17 +1,236 @@
+      if(notEmpty(campaign_name), campaign_name, 'organic') AS campaign,
+      if(notEmpty(source_name), source_name, 'organic') AS source,
+      count() AS conversion_0
+  
+  FROM
+      (
+  SELECT
+          person_id,
+          if(notEmpty(conversion_campaign), conversion_campaign, if(notEmpty(fallback_campaign), fallback_campaign, '')) AS campaign_name,
+          if(notEmpty(conversion_source), conversion_source, if(notEmpty(fallback_source), fallback_source, '')) AS source_name,
+          1 AS conversion_value
+      
+  FROM
+          (
+  SELECT
+              person_id,
+              conversion_timestamps[i] AS conversion_time,
+              conversion_math_values[i] AS conversion_math_value,
+              conversion_campaigns[i] AS conversion_campaign,
+              conversion_sources[i] AS conversion_source,
+              arrayMax(arrayFilter(x -> and(lessOrEquals(x, conversion_timestamps[i]), greaterOrEquals(x, minus(conversion_timestamps[i], 15552000))), utm_timestamps)) AS last_utm_timestamp,
+              if(isNotNull(last_utm_timestamp), utm_campaigns[indexOf(utm_timestamps, last_utm_timestamp)], '') AS fallback_campaign,
+              if(isNotNull(last_utm_timestamp), utm_sources[indexOf(utm_timestamps, last_utm_timestamp)], '') AS fallback_source
+          
+  FROM
+              (
+  SELECT
+                  events.person_id,
+                  arrayFilter(x -> greater(x, 0), groupArray(if(equals(events.event, 'user signed up'), toUnixTimestamp(events.timestamp), 0))) AS conversion_timestamps,
+                  arrayFilter(x -> greater(x, 0), groupArray(if(equals(events.event, 'user signed up'), toFloat(1), 0))) AS conversion_math_values,
+                  arrayFilter(x -> notEmpty(toString(x)), groupArray(if(equals(events.event, 'user signed up'), toString(ifNull(events.properties.utm_campaign, '')), ''))) AS conversion_campaigns,
+                  arrayFilter(x -> notEmpty(toString(x)), groupArray(if(equals(events.event, 'user signed up'), toString(ifNull(events.properties.utm_source, '')), ''))) AS conversion_sources,
+                  arrayFilter(x -> greater(x, 0), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toUnixTimestamp(events.timestamp), 0))) AS utm_timestamps,
+                  arrayFilter(x -> notEmpty(x), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toString(ifNull(events.properties.utm_campaign, '')), ''))) AS utm_campaigns,
+                  arrayFilter(x -> notEmpty(x), groupArray(if(and(equals(events.event, '$pageview'), notEmpty(toString(ifNull(events.properties.utm_campaign, ''))), notEmpty(toString(ifNull(events.properties.utm_source, '')))), toString(ifNull(events.properties.utm_source, '')), ''))) AS utm_sources
```

Comment:
If you can find a way to use a single array function here, I think it could speed up the query.